runtime/src/main/java/org/corfudb/runtime/collections/CorfuQueue.java

```diff
@@ -0,0 +1,207 @@
+        put(id, e);
+        return id;
+    }
+
+    /** CorfuQueueRecord encapsulates each entry enqueued into CorfuQueue with its unique ID.
+     * It is a read-only type returned by the entryList() method.
+     * The ID returned here can be used for both point get()s as well as remove() operations
+     * on this Queue.
+     *
+     * @param <E>
+     */
+    public static class CorfuQueueRecord<E> {
+        /**
+         * This ID represents the entry and its order in the Queue.
+         * This implies that it is unique and comparable with other IDs
+         * returned from CorfuQueue methods with respect to its enqueue order.
+         */
+        @Getter
+        final UUID ID;
+
+        @Getter
+        final E entry;
+
+        CorfuQueueRecord(UUID id, E entry) {
+            this.ID = id;
+            this.entry = entry;
+        }
+    }
+
+    /**
+     * Returns a List of CorfuQueueRecords sorted by the order in which the enqueue materialized.
+     *
+     * {@inheritDoc}
+     *
+     * <p>This function currently does not return a view like the java.util implementation,
+     * and changes to the entryList will *not* be reflected in the map. </p>
+     *
+     * @param maxEntires - Limit the number of entries returned from start of the queue
+     *                   -1 implies all entries
+     * @return List of Entries sorted by their enqueue order
+     */
+    public List<CorfuQueueRecord<E>> entryList(int maxEntires) {
+        if (maxEntires == -1) {
+            maxEntires = Integer.MAX_VALUE;
+        }
+        List<CorfuQueueRecord<E>> copy = new ArrayList<>(Math.min(mainMap.size(), maxEntires));
+        for (Map.Entry<UUID, E> entry : mainMap.entrySet()) {
+            copy.add(new CorfuQueueRecord<>(entry.getKey(),
+                    entry.getValue()));
+            if (--maxEntires == 0) {
+                break;
+            }
+        }
+        return copy;
+    }
+
+    @Accessor
+    public boolean isEmpty() {
+        return mainMap.isEmpty();
+    }
+
+    /** {@inheritDoc} */
+    @Accessor
+    public boolean containsKey(@ConflictParameter UUID key) {
+        return mainMap.containsKey(key);
+    }
+
+    /** {@inheritDoc} */
+    @Accessor
+    public E get(@ConflictParameter UUID key) {
+        return mainMap.get(key);
+    }
+
+    /**
+     * Removes a specific element identified by the ID returned via entryList()'s CorfuQueueRecord.
+     *
+     * {@inheritDoc}
+     *
+     * @throws NoSuchElementException if this queue did not contain this element
+     * @return The entry that was successfully removed
+     */
+    @MutatorAccessor(name = "remove", undoFunction = "undoRemove",
+                                undoRecordFunction = "undoRemoveRecord")
+    @SuppressWarnings("unchecked")
+    public E remove(@ConflictParameter UUID key) {
+        E previous =  mainMap.remove(key);
+        return previous;
+    }
+
+    /**
+     * Erase all entries from the Queue.
```

Comment:
Can we rephrase the 'erase'? Maybe remove.

Reply:
Done.